experimental/dds/tree2/src/util/utils.ts

```diff
@@ -266,6 +271,46 @@ export function assertNonNegativeSafeInteger(index: number) {
  */
 export type Assume<TInput, TAssumeToBe> = TInput extends TAssumeToBe ? TInput : TAssumeToBe;
 
+/**
+ * The counter used to generate deterministic stable ids for testing purposes.
+ */
+let deterministicStableIdCount: number | undefined;
+
+/**
+ * This function is used to generate deterministic stable ids for testing purposes.
+ * @param f - A function that will be called and if a stable id is needed inside it a deterministic one will be used.
+ *
+ * @remarks
+ * Only use this function for testing purposes.
+ *
+ * @example
+ * ```ts
+ * const summary = useDeterministicStableId(() => tree.summarize());
```

Comment:
not the best example--this implementation won't work for asynchronous functions, which `summarize` is. Kinda indifferent on whether it's better to allow async functions or not. I'd be ok with either of:

1. document them as unsupported and use a different example here
2. support them by making `useDeterministicStableId` asynchronous as well as `f`, then awaiting its result before resetting `deterministicStableIdCount`.

Reply:
`deterministicStableIdCount` is a shared, module-level variable, so if `useDeterministicStableId` is async then multiple concurrent uses of this function could interleave and clobber each other.

Reply:
oof yeah, good point. if we wanted async then we'd want to assert there isn't any double-calls (still doable, but does make me think banning async might be the better option)

Reply:
It's already broken if `useDeterministicStableId` is called inside the `f` of another call to `useDeterministicStableId`. We should add an assert for that too (assert that `deterministicStableIdCount === undefined` when function starts). Or, you can handle it gracefully by setting aside the starting value and restoring the starting value at the end of the function, rather than always setting to `undefined`. That's probably easy enough to do that it's worth it.